### PR TITLE
Convert legacy font prefixes into meaningful font prefixes

### PR DESCRIPTION
**Pull request by roscale**
* Legacy prefixes had no reason to exists anymore. They have been replaced by more meaningful value (yet, limited to 5 characters). The new prefixes no longer contain hardcoded font sizes because it has been possible for some time to display text of any size.
* It is now possible to create custom fonts.
* Default sizes have been removed from font objects, because font size is specified in `TextDrawing` objects.
* There are now 4 predefined fonts: Roboto Mono normal, bold, italic, and italic bold.

### CODE REVIEW

**Comment by roscale** [comment]
See https://github.com/getyourway/gyw-ardent/issues/121#issuecomment-2084752913

This is why I stopped here because I need your input.

File: lib/src/drawings.dart [hunk 1]
```diff
@@ -158,8 +154,8 @@ class TextDrawing extends GYWDrawing {
     controlBytes.add(int16Bytes(left));
     controlBytes.add(int16Bytes(top));
 
-    controlBytes.add(utf8.encode(font?.prefix ?? "NUL"));
-    controlBytes.add(int8Bytes(size ?? 0));
+    controlBytes.add(utf8.encode(font.filename.padRight(5, "\u0000")));
```
Comment (TheToto42) [inline lib/src/drawings.dart:157]: I'm scared of allowing padded font, can't we use an assert in the constructor or something similar ? isn't it safer and easier to handle

File: lib/src/drawings.dart [hunk 2]
```diff
@@ -210,24 +206,18 @@ class TextDrawing extends GYWDrawing {
...
-          )
-          .font;
-    } on StateError {
-      font = null;
-    }
+    final font = GYWFonts.values
+        .firstWhere(
+          (e) => e.font.name == data["font"],
+        )
+        .font;
```
Comment (TheToto42) [inline lib/src/drawings.dart:213]: Either add a default value, either create something similar as with custom icons

File: lib/src/drawings.dart [hunk 3]
```diff
@@ -241,9 +231,7 @@ class TextDrawing extends GYWDrawing {
       "left": left,
       "top": top,
       "data": text,
-      // Deprecated: "text" key will be deprecated in future version
-      "text": text,
-      if (font != null) "font": font!.prefix,
+      "font": font.name,
```
Comment (TheToto42) [inline lib/src/drawings.dart:234]:
```suggestion
      "font": font.filename,
```

File: lib/src/fonts.dart [hunk 4]
```diff
@@ -6,96 +6,48 @@ class GYWFont {
...
-      height: 33,
-      bold: true,
-    ),
-  ),
-
-  /// For common texts with a medium significance.
-  large(
+  /// Roboto Mono Normal
+  robotoMonoNormal(
     GYWFont(
```
Comment (TheToto42) [inline lib/src/fonts.dart:40]: Shouldn't we keep `bold` and `italic` parameters ? and have something like `pathSvg` that returns the font to use in Flutter preview ?

File: test/drawing_test.dart
Comment (TheToto42) [inline test/drawing_test.dart:1]: Why did you change what kind of tests are done ?

File: lib/src/drawings.dart [hunk 5]
```diff
@@ -85,8 +83,8 @@ class TextDrawing extends GYWDrawing {
   /// Creates a text element.
   const TextDrawing({
     required this.text,
-    this.font,
-    this.size,
+    required this.font,
+    required this.size,
```
[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart:87]: Sure.

File: lib/src/drawings.dart [hunk 6]
```diff
@@ -96,8 +94,7 @@ class TextDrawing extends GYWDrawing {
 
   @override
   List<GYWBtCommand> toCommands() {
-    final int fontSize = size ?? font?.size ?? GYWFonts.small.font.size;
-    final int charHeight = (fontSize * 1.33).ceil();
+    final int charHeight = (size * 1.33).ceil();
```
[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart:97]: I don't think I want to put char height as a property in font. This value doesn't make sense when we will add non-monospaced fonts. Besides, why would you want it to make this property part of the public interface?

If, for some reason, you want to give the user monospace properties about fonts that you provide, then lets have 2 types of fonts: monospace and non-monospace (idk if this is what they are called). This way we don't have to deprecate anything when we add non-monospace fonts.

File: lib/src/drawings.dart (hunk 1 above)
Comment (roscale) [reply lib/src/drawings.dart:157]: It's not necessarily safer because I always add a null terminator when I parse the command, so it's the same. But sure. I'm impartial.

File: test/drawing_test.dart
Comment (roscale) [reply test/drawing_test.dart:1]: Tests with and without fonts can be merged together because font is now a required property.
So 4 tests become 2.

File: lib/src/drawings.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart]: Should I also default the font to roboto mono normal?

[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart]: Acutally I can't. 

`The property 'font' can't be accessed on the type 'GYWFonts' in a constant expression.`

Too bad. It will stay required then.

**Comment by roscale** [comment]
I've put some comments.

File: lib/src/drawings.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheToto42) [reply lib/src/drawings.dart:97]: I see what you mean, but for now, there are no non-monospaced (idk either if it's the correct name) fonts. So, in the current version, we can assume that all fonts are monospaced.

`charHeight` will probably have meaning for non-monospaced fonts too. Perhaps, it will refer to maximum char height, but we will still need this value to compute wrapping and features like this. I agree that with `charWidth` it is less obvious, but it may also refers to the maximum width of a character. I personally use these value a lot when building a screen with variable size content in order to place an icon at the right location for example.

File: lib/src/drawings.dart (hunk 1 above)
Comment (TheToto42) [reply lib/src/drawings.dart]: > It's not necessarily safer because I always add a null terminator when I parse the command,

That's a good a safe way to manage it on the firmware side. Let's still add the `assert` here to discourage the user to do so

File: lib/src/fonts.dart (hunk 4 above)
Comment (TheToto42) [reply lib/src/fonts.dart:52]: > Sure. Is it just because you want to offer more information to the user just in case?

I want the user to easily 'convert' his text in a preview

> `pathTtf` you meant probably. Or should we just name it `path`?

You can name it  `path`. But in reality, i'm thinking that when we use the font in a `TextStyle` object, we never use the ttf value. So I'm not sure if it is really relevant to have this. The goal is to but the value in the `TextStyle` object, what do you suggest ?

**Comment by TheToto42** [comment]
> Is this sort of message helpful, or do you receive notifications for all the replies above?

I had the notification that some messages were available on the PR, don't worry

File: lib/src/fonts.dart [hunk 7]
```diff
@@ -3,99 +3,76 @@ import "package:flutter/material.dart";
 /// The text fonts supported on aRdent smart glasses
 @immutable
 class GYWFont {
-  /// The name of the font
+  /// The name of the font.
   final String name;
 
-  /// The internal prefix of the font
-  final String prefix;
+  /// The font filename on the GYW device.
```
Comment (TheToto42) [inline lib/src/fonts.dart:9]:
```suggestion
  /// The font filename on the GYW device (5 characters-long and no type extension).
```

File: lib/src/drawings.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart:97]: I see. I didn't foresee this use case. Indeed, if you want to compute layout and have other drawings' positions depend on its siblings, than ok, the best we can do right now it to provide a max width and height.

[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart]: What do you think about the latest commit?

1.333 is actually just a constant because it's the size of a pixel per per point.
0.6 (charWidth) can be derived from the aspect ratio of the font. All fonts, monospaced and non-monospaced have an aspect ratio, and ours is 0.45.

File: lib/src/drawings.dart (hunk 1 above)
Comment (TheToto42) [reply lib/src/drawings.dart]: yes, it's fine. Just add the error message related to the assert. I think there is a rule in `flutter_lints` that prevents to write assert without the associate message, can you enable it ?

File: lib/src/fonts.dart (hunk 4 above)
Comment (TheToto42) [reply lib/src/fonts.dart:62]: Perfect, font family seems perfect. With that, bold and italic will make more sense as we may have a font family that is linked to several `GYWFont`

File: lib/src/drawings.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (TheToto42) [reply lib/src/drawings.dart]: Let's have 1.333 hardcoded like you did with the calculation, but it shouldn't be in the public API.

For the aspect ratio, how do you make the translation from the aspect ratio to char width ? Is it 0.45 * 1.333 ? I feel like charWidth is easier to understand and to use, but I see the more general idea behind it. In non)monospaced font, the aspect ratio is fixed ?

File: lib/src/fonts.dart [hunk 8]
```diff
@@ -1,101 +1,91 @@
...
     required this.name,
-    required this.prefix,
-    required this.size,
-    required this.width,
-    required this.height,
-    required this.bold,
-  });
+    required this.filename,
+    required this.pathTtf,
+    required this.aspectRatio,
```
Comment (TheToto42) [inline lib/src/fonts.dart:35]: You can give a default value to these. The goal is to be able to create a `GYWFont` object based only on the filename, in a `TextDrawing.fromJson` method for example.

File: lib/src/fonts.dart [hunk 9]
```diff
@@ -1,101 +1,88 @@
...
-    required this.name,
-    required this.prefix,
-    required this.size,
-    required this.width,
-    required this.height,
-    required this.bold,
-  });
+    required this.filename,
+    this.name = "Custom Font",
+    this.fontFamily = "Custom Font",
```
Comment (TheToto42) [inline lib/src/fonts.dart:34]: I've thought at something regarding these two. What do you think of having them nullable, and then null by default ?

Comment (roscale) [reply lib/src/fonts.dart:34]: Yes, that's better. Optionality should be conveyed throurgh nullability.

File: lib/src/drawings.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart]: @TheToto42 I don't like that Dart doesn't allow you to use enum values in const contexts. Should I make font nullable and set if it's null, choose roboto mono by default?

[the remark this answers is not in the document]
Comment (TheToto42) [reply lib/src/drawings.dart]: Yes, you can do that

[the remark this answers is not in the document]
Comment (roscale) [reply lib/src/drawings.dart]: I solved this in a different way. If I let font be nullable, I would have to pay attention to make sure that null == roboto mono in the equality operator and the methods that use this font, and I didn't quite like that. I decided to work around the issue (https://github.com/dart-lang/language/issues/2374) by just defining a constant with the font, and the constructor can reference this constant instead of `GYWFonts.robotoMono.font`.

https://github.com/getyourway/flutter_gyw/blob/4b209fa0324cf2eee2247bdb94ef3cf2ff184947/lib/src/fonts.dart#L58-L70

Is this solution ok to you?

**Comment by roscale** [comment]
I also defined the italics font in this PR. Should I remove it and make another PR in the future when we need it, or should I leave it and make a PR that adds the file to the filesystem on the glasses?

**Review by TheToto42 — APPROVED** [review]
It's annoying that Flutter forces us to have the workaround, but we'll have to live with it

**Comment by roscale** [comment]
I also put another question here, I don't think you've seen it: https://github.com/getyourway/flutter_gyw/pull/68#issuecomment-2096260151

**Comment by TheToto42** [comment]
> I also put another question here, I don't think you've seen it: https://github.com/getyourway/flutter_gyw/pull/68#issuecomment-2096260151

Indeed, I did not see it. You can add the fonts here, and create a PR on the firmware repo to add the italic fonts (italic and italic bold)

**Review by TheToto42 — APPROVED** [review]
